hubtools/src/lib.rs

```diff
@@ -602,4 +662,125 @@ impl RawHubrisArchive {
+        let cmpa_bytes = self.extract_file("img/CMPA.bin")?;
+        if cmpa_bytes.len() != 512 {
+            return Err(Error::BadCMPASize(cmpa_bytes.len()));
+        }
+        let mut cmpa_array = [0u8; 512];
+        cmpa_array.copy_from_slice(&cmpa_bytes);
+        let cmpa = lpc55_areas::CMPAPage::from_bytes(&cmpa_array)?;
+
+        let cfpa_bytes = self.extract_file("img/CFPA.bin")?;
+        if cfpa_bytes.len() != 512 {
+            return Err(Error::BadCFPASize(cfpa_bytes.len()));
+        }
+        let mut cfpa_array = [0u8; 512];
+        cfpa_array.copy_from_slice(&cfpa_bytes);
+        let cfpa = lpc55_areas::CFPAPage::from_bytes(&cfpa_array)?;
+
+        lpc55_sign::verify::init_verify_logger(verbose);
+        lpc55_sign::verify::verify_image(&self.image.data, cmpa, cfpa)?;
+
+        Ok(())
+    }
+
+    /// Signs the given image with a chain of one-or-more certificates
+    ///
+    /// This modifies local data in memory; call `self.overwrite` to persist
+    /// changes back to the archive on disk.
+    pub fn sign(
+        &mut self,
+        signing_certs: Vec<Vec<u8>>,
+        root_certs: Vec<Vec<u8>>,
+        private_key: &str,
+        execution_address: u32,
+    ) -> Result<(), Error> {
+        let manifest = self.extract_file("app.toml")?;
+        let manifest: toml::Value = toml::from_str(
+            std::str::from_utf8(&manifest).map_err(Error::BadManifest)?,
+        )
+        .map_err(Error::BadToml)?;
+        let chip = manifest
+            .as_table()
+            .ok_or(Error::BadTomlType)?
+            .get("chip")
+            .ok_or(Error::BadTomlType)?
+            .as_str()
+            .ok_or(Error::BadTomlType)?
+            .to_owned();
+
+        if !chip.contains("lpc55") {
+            return Err(Error::WrongChip(chip));
+        }
+
+        self.image.sign(
+            signing_certs,
+            root_certs,
+            private_key,
+            execution_address,
+        )
+    }
+
+    /// Adds `img/CMPA.bin` to the archive, generated based on a DICE
+    /// configuration and set of root certificates.
+    ///
+    /// This modifies local data in memory; call `self.overwrite` to persist
+    /// changes back to the archive on disk.
+    pub fn set_cmpa(
+        &mut self,
+        dice: lpc55_sign::signed_image::DiceArgs,
+        enable_secure_boot: bool,
+        debug: lpc55_areas::DebugSettings,
+        default_isp: lpc55_areas::DefaultIsp,
+        speed: lpc55_areas::BootSpeed,
+        boot_error_pin: lpc55_areas::BootErrorPin,
+        root_certs: Vec<Vec<u8>>,
+    ) -> Result<(), Error> {
+        let rkth = lpc55_sign::signed_image::root_key_table_hash(root_certs)?;
+        let cmpa = lpc55_sign::signed_image::generate_cmpa(
+            dice,
+            enable_secure_boot,
+            debug,
+            default_isp,
+            speed,
+            boot_error_pin,
+            rkth,
+        )?;
+        const CMPA_FILE: &str = "img/CMPA.bin";
```

Comment:
(Extracting this as a constant is good, but I notice this constant appears in several places... move it to module scope?)

Reply:
good idea, done